### DISCUSSION TITLE
How to redirect a url to another language version which is only supported in different location? to search websites by preferred language but open a search result in another language which in my case is only provisioned in different location.

### DISCUSSION THREAD

**Discussion by sunnyintoronto** [mshibanami/redirect-web#56] [Questions] [answered]
Example of original url
https://support.microsoft.com/en-us/outlook
https://support.microsoft.com/zh-hk/outlook

**Comment by mshibanami** [comment]
Hey @sunnyintoronto, this should work:

- Redirect From: `https://support.microsoft.com/en-us/*` (Wildcard)
- Redirect To: `https://support.microsoft.com/zh-hk/$1`

Wildcard's `*` is captured and you can reference it with `$1`.
See more details here: https://mshibanami.github.io/redirect-web/rule-settings

**Reply by zluckytraveler** [reply]
Is there a way to try and apply this method for all websites where there is language specified anywhere within the url that isn’t in `en` or `en-us`?

**Reply by mshibanami** [reply]
Hi @zluckytraveler, I think it depends on what you mean by "all websites."
If you intend to redirect only those websites that you know support your language, you can use this rule:

- Redirect From: `(https?://(?:support.microsoft.com|example.com))/en(?:-us)?/(.*)` (Regular Expression)  
    - You can specify multiple domains using the `|` (pipe) symbol.
- Redirect To: `$1/zh-hk/$2`

On the other hand, if you truly want to apply this rule to every website, it would look like this:

- Redirect From: `(https?://[^/]*)/en(?:-us)?/(.*)` (Regular Expression)
- Redirect To: `$1/zh-hk/$2`

However, I'm not sure if you want to do this, as many websites only support English. Such websites might redirect you back to the English page, potentially causing an infinite loop.

I hope it helps!

**Reply by zluckytraveler** [reply]
> Hi @zluckytraveler, I think it depends on what you mean by "all websites."
> 
> If you intend to redirect only those websites that you know support your language, you can use this rule:
> 
> 
> 
> - Redirect From: `(https?://(?:support.microsoft.com|example.com))/en(?:-us)?/(.*)` (Regular Expression)  
> 
>     - You can specify multiple domains using the `|` (pipe) symbol.
> 
> - Redirect To: `$1/zh-hk/$2`
> 
> 
> 
> On the other hand, if you truly want to apply this rule to every website, it would look like this:
> 
> 
> 
> - Redirect From: `(https?://[^/]*)/en(?:-us)?/(.*)` (Regular Expression)
> 
> - Redirect To: `$1/zh-hk/$2`
> 
> 
> 
> However, I'm not sure if you want to do this, as many websites only support English. Such websites might redirect you back to the English page, potentially causing an infinite loop.
> 
> 
> 
> I hope it helps!

It looks like what you had mentioned is doing the opposite of what I was intending. 

It appears the redirect rules are trying to change all websites that **are** in English to Chinese.

What I'm referring to is, for all websites where there is a specified language in the url that is **not** in `en` or `en-us` to change it to be so.

Not sure if it's possible as I'm not good with regexes. It would need to be able to determine any part of the url that is specifically in every [ISO 639-1](https://www.wikiwand.com/en/articles/ISO_639-1) and [ISO 3166-1 alpha-2](https://www.wikiwand.com/en/articles/ISO_3166-1_alpha-2) language code except for English (`en` or `en-us`).

Using Spanish as a only single _Redirect From_ language example only. Typically most urls have languages designated as `.es.`, `/es/`, `.es/`, `/es.` (same concept for `es-mx`,`es-es`, etc), and the placement of it within the url won't be in the same spot for all websites.

I know these are some hurdles, and not sure if it's possible, but if so think it'd be an absolutely great addition to the library and in general, if it could work.

**Reply by mshibanami** [reply]
@zluckytraveler Oh, I see. Sorry for the misunderstanding, and thank you for the clarification!

In that case, I think you can start with this rule:

- Redirect From: `(https?://[^/]*)/(?:es|ja)((?:[_-](?:ES|es|MX|mx|JP|jp))?)/(.*)`
- Redirect To: `$1/en$2/$3`
- Capture Group Processing:
    - 1st Processing
        - Group: `$2`
        - Process: `Replace Occurrences`
        - Target: `[A-Z][A-Z]` (Regular Expression)
        - Replacement: `US`
    - 2nd Processing
        - Group: `$2`
        - Process: `Replace Occurrences`
        - Target: `[a-z][a-z]` (Regular Expression)
        - Replacement: `us`

📷 Screenshot: https://github.com/user-attachments/assets/3f50d03c-e883-4412-9d79-d9acdfd3a2b1

This rule supports Spanish and Japanese, with or without country codes, and redirects like this:

- https://example.com/ja/outlook → https://example.com/en/outlook
- https://example.com/ja-JP/outlook → https://example.com/en-US/outlook
- https://example.com/es_ES/outlook → https://example.com/en_US/outlook
- https://example.com/es-es/outlook → https://example.com/en-us/outlook

I understand this is a simple solution and may not cover all cases (e.g., it only handles `https://{domain}/{language}/...` format and doesn't properly address `es-mx` or `es-es`). You may need to specify Excluded URL Patterns or create multiple rules to handle edge cases. Still, I think it's a good starting point.

Feel free to let me know if you run into any issues!

By the way, that sounds like an interesting use case. Maybe I'll add a related rule to the Library section. 😄

**Reply by zluckytraveler** [reply]
This is great, thanks! 

I'll definitely give this a shot, add some language codes myself too, and let you know if I run into any issues!

On a side note, I don't think Uppercase letters are quite needed, since they are always treated the same as lowercase. Unless for some reason a website is hosted on a server with a case-sensitive file system, is the only exception I could think of.

I personally doubt that there would be huge use for languages in every country code. However, using something like the top 3 countries for major languages would be a great place to start. 

If something like this is added later and a way for the placement of language code could be anywhere within the URL. This would really make this Redirect Rule standout and work on nearly every website, with some very extreme edge cases.

I really hope to see this redirect rule progress and make its way to the Library.

**Reply by sunnyintoronto** [reply]
I need RegEx for iPhone iOS 18.4.1 which doesn't accept RegEx in Replacement. Thanks for your help in advance!

**Reply by mshibanami** [reply]
Hi @sunnyintoronto, what does your rule actually look like? Would you mind sending me a video or a screenshot of your issue?

**Comment by sunnyintoronto** [comment]
Thanks my RegEx skills is a beginner. According to https://www.reddit.com/r/shortcuts/s/Xm53OZXErQTo replace multiple LFs with single LF, iPhone iOS 18.4.1 Shortcut App Action Replace With field does not support \n with (RegEx enabled) ; I have to useReplace Text - "\n\n" with "" in [input], Regular Expression ONHere is what I researched using AI # iPhone (iOS) RegEx LimitationsWhen working with regular expressions on iPhone/iOS, there are some limitations and implementation details to be aware of:## Core Limitations1. **NSRegularExpression Implementation**:   - iOS uses the NSRegularExpression class which is based on ICU (International Components for Unicode) library   - Doesn't support some advanced regex features found in PCRE or Perl2. **Lookbehind Limitations**:   - Only fixed-length lookbehind is supported (no variable-length lookbehind)   - Positive lookbehind (`(?<=...)`) is supported   - Negative lookbehind (`(?<!...)`) is supported3. **Unicode Support**:   - Full Unicode support is available but may have performance implications   - Some Unicode property escapes might not work as expected## Common Issues- No support for recursive patterns or balancing groups- Limited support for some advanced backtracking control verbs- Atomic grouping (`(?>...)`) is supported but possessive quantifiers might not be- Named capture groups are supported in newer iOS versions (iOS 11+)## Performance Considerations- Complex regex patterns can cause performance issues on mobile devices- Very long input strings may lead to noticeable processing delays## WorkaroundsFor complex regex needs:1. Break the problem into multiple simpler regex operations2. Consider doing some processing in native Swift/Obj-C code3. For web content, consider using JavaScript regex instead through WKWebView## Version DifferencesNewer iOS versions have improved regex support, so always check the minimum iOS version your app supports when using advanced features.On 29 Apr 2025, at 03:59, Manabu Nakazawa ***@***.***> wrote:﻿
Hi @sunnyintoronto, what does your rule actually look like? Would you mind sending me a video or a screenshot of your issue?

**Reply by mshibanami** [reply, marked as answer]
@sunnyintoronto Sorry, I'm not sure I understand you correctly. Are you asking how to use the Shortcuts app? The link you included (https://www.reddit.com/r/shortcuts/s/Xm53OZXErQTo) doesn't seem to take me to the right place.

If you prefer email communication, feel free to reach out to me at: mshibanami+redirectweb@gmail.com

**Reply by sunnyintoronto** [reply]
Yes. I need to use iPhone Browser to go to website from English to another language for user manual webpage because I am more proficient in English when searching but need user manual in different language.

**Reply by mshibanami** [reply]
@sunnyintoronto I'm not very familiar with the Shortcuts app. But in that case, I think you can try [the rule I already suggested](https://github.com/mshibanami/redirect-web/discussions/56#discussioncomment-10789013) for Redirect Web instead of using the Shortcuts app. Please note that the Shortcuts app is Apple's official app and Redirect Web is only for Safari (iPhone Browser).

If you still have any questions about setting up the rule, could you please send me a video or a screenshot of the issue?

**Reply by zluckytraveler** [reply]
@sunnyintoronto To be honest it sounds like you’d benefit more specifically from a translation app that integrates with safari extensions. Finding English websites in foreign languages isn’t always going to be available. 

You could try using [Immersive](https://apps.apple.com/us/app/immersive-translate-for-safari/id6447957425). I’ve used it in the past myself, it’s 100% free and it’s pretty darn good.

@mshibanami I hope this ok to mention another app to help them out. If not you can just remove, and I apologize.
